File: src/DlgMain_funcs.cpp
#include "DlgMain.h"
#include "id3v2/Tag.h"
#include "../res/resource.h"
using std::optional, std::vector, std::wstring, std::wstring_view;

void DlgMain::addMp3sToList(const vector<wstring>& mp3s) const
{
	lib::ListView lv{this, LST_FILES};

	vector<wstring> nonMp3s; // keep track of files that aren't MP3
	for (const wstring& mp3 : mp3s) {
		if (!lib::path::hasExtension(mp3, L"mp3") && !lib::path::isDir(mp3))
			nonMp3s.emplace_back(mp3);
	}
	if (!nonMp3s.empty()) {
		wstring buf = lib::str::newReserved(22 * nonMp3s.size()); // arbitrary
		buf = L"Non-MP3 file(s):";
		for (const wstring& mp3 : nonMp3s) {
			buf.append(L"\n");
			buf.append(mp3);
		}
		dlg.msgBox(L"Non-MP3 file(s)", {}, buf, TDCBF_OK_BUTTON, TD_ERROR_ICON);
		return; // do not continue; no files will be added
	}

	for (const wstring& mp3 : mp3s) { // all files are MP3, let's add them
		if (lib::path::isDir(mp3)) {
			for (const wstring& f : lib::path::dirList(mp3 + L"\\*.mp3")) // search only 1 level deep
				addOneMp3ToList(f);
		} else {
			addOneMp3ToList(mp3);
		}
	}
	updateNumFiles(lv.items.count());
	lv.columns[0].setWidthToFill();
	sortList();
}

void DlgMain::addOneMp3ToList(wstring_view mp3) const
{
	lib::ListView lv{this, LST_FILES};
	int idxItem = -1;
	if (auto curItem = lv.items.find(mp3); curItem.has_value()) { // MP3 already in list
		auto pTagCurrent = curItem.value().data<id3::Tag*>();
		delete pTagCurrent;
		idxItem = curItem.value().index();
	} else { // MP3 not in list yet
		auto newItem = lv.items.add(mp3); // insert new
		idxItem = newItem.index();
	}

	auto item = lv.items[idxItem];
	id3::Tag* pTag = nullptr;
	try {
		pTag = new id3::Tag{mp3}; // store pointer to Tag in item
	} catch (const std::exception& e) {
		item.remove();
		auto msg = lib::str::fmt(L"%s\n\n%s", mp3, lib::str::toWide(e.what()));
		dlg.msgBox(L"Tag parsing error", {}, msg, TDCBF_OK_BUTTON, TD_ERROR_ICON);
		return;
	}
	item.setData(pTag);
	renderMp3ListItem(item);
}

void DlgMain::renderMp3ListItem(lib::ListView::Item item) const
{
	auto pTag = item.data<const id3::Tag*>();

	auto renderSimple = [item, pTag](wstring_view name4, UINT col) {
		if (optional<const id3::Frame*> frame = pTag->frameByName4(name4); frame.has_value()) {
			auto pFrameText = frame.value()->dataAs<id3::Frame::Text>();
			item.setText(pFrameText->text, col);
		} else {
			item.setText(L"", col); // removed frames need to have their text erased
		}
	};

	item.setText(std::to_wstring(pTag->padding), 1);
	if (auto pic = pTag->frameByName4(L"APIC"); pic.has_value()) {
		item.setText(L"\u2713", 2); // checkmark
	} else {
		item.setText(L"", 2);
	}
	item.setText(pTag->replayGainStatus(), 3);
	renderSimple(L"TPE1", 4);
	renderSimple(L"TYER", 5);
	renderSimple(L"TALB", 6);
	renderSimple(L"TRCK", 7);
	renderSimple(L"TIT2", 8);
	renderSimple(L"TCON", 9);
	renderSimple(L"TPE3", 10);
	renderSimple(L"TCOM", 11);
	renderSimple(L"TEXT", 12);
	renderSimple(L"TOPE", 13);
	if (auto comm = pTag->frameByName4(L"COMM"); comm.has_value()) {
		auto pFrameComm = comm.value()->dataAs<id3::Frame::Comment>();
		item.setText(pFrameComm->text, 14);
	} else {
		item.setText(L"", 14);
	}
}

void DlgMain::updateNumFiles(UINT numFiles) const
{
	setText(lib::str::fmt(L"ID3 Fit (%d/%d)",
		lib::ListView{this, LST_FILES}.items.countSelected(), numFiles));
}

void DlgMain::sortList() const
{
	using lib::ListView;
	ListView{this, LST_FILES}.items.sort([this](ListView::Item a, ListView::Item b) -> int {
		int cmp = 0;
		if (_sort.col == 0 || _sort.col == 1) {
			auto pTagA = a.data<const id3::Tag*>();
			auto pTagB = b.data<const id3::Tag*>();
			if (_sort.col == 0) { // by path
				cmp = lib::str::cmpI(pTagA->path, pTagB->path);
			} else { // by padding size
				cmp = pTagA->padding - pTagB->padding;
			}
		} else { // by column text
			cmp = lib::str::cmpI(a.text(_sort.col), b.text(_sort.col));
		}
		return _sort.asc ? cmp : -cmp;
	});
}

void DlgMain::saveSelected() const
{
	struct Fail final {
		wstring file;
		std::string reason;
	};

	auto selItems = lib::ListView{this, LST_FILES}.items.selected();
	auto faileds = lib::vec::newReserved<Fail>(selItems.size());
	size_t numSaved = 0;
	auto t0 = lib::TimeCount::Immediately();

	for (auto&& item : selItems) {
		auto pTag = item.data<id3::Tag*>();
		try {
			pTag->saveToFile();
			++numSaved;
		} catch (const std::exception& e) {
			faileds.push_back({.file = pTag->path, .reason = e.what()});
		}
	}

	auto t1 = t0.now();

	auto mainMsg = lib::str::fmt(L"%d tag(s) saved in %02d.%03d.", numSaved, t1.sec, t1.ms);
	if (faileds.empty()) {
		dlg.msgBox(L"Tag(s) saved", {}, mainMsg, TDCBF_OK_BUTTON, TD_INFORMATION_ICON);
	} else {
		mainMsg += L"\n\n" + std::to_wstring(faileds.size()) + L" file(s) failed:";
		for (auto&& failed : faileds)
			mainMsg += L"\n\n" + failed.file + L"\n" + lib::str::toWide(failed.reason);
		dlg.msgBox(L"Tag(s) saved with errors", {}, mainMsg, TDCBF_OK_BUTTON, TD_ERROR_ICON);
	}
}

File: dlg/dlgmain/DlgMainFuncs.go
//go:build windows

package dlgmain

import (
	"fmt"
	"id3fit/dlg/dlgedit"
	"id3fit/id3v2"
	"strings"

	"github.com/rodrigocfd/windigo/co"
	"github.com/rodrigocfd/windigo/ui"
	"github.com/rodrigocfd/windigo/win"
	"github.com/rodrigocfd/windigo/wstr"
	"github.com/rodrigocfd/xslices"
)

func (me *DlgMain) setWaitState(set bool) {
	if set {
		me.wnd.Hwnd().SetWindowText("Working...")
		me.lstFiles.Hwnd().EnableWindow(false)
		me.isWaiting = true
	} else {
		me.updateTitlebarCount()
		me.lstFiles.Hwnd().EnableWindow(true)
		me.isWaiting = false // read in WM_SETCURSOR
		cPos, _ := win.GetCursorPos()
		win.SetCursorPos(int(cPos.X), int(cPos.Y)) // force cursor redraw
	}
}

func (me *DlgMain) updateTitlebarCount() {
	nFiles := me.lstFiles.Items.Count()
	nSel := me.lstFiles.Items.SelectedCount()
	me.wnd.Hwnd().SetWindowText(fmt.Sprintf("ID3 Fit (%d/%d)", nSel, nFiles))
}

func (me *DlgMain) addMp3sToListAsync(incomingPaths []string) {
	me.setWaitState(true)
	go func() {
		allPaths := make([]string, 0, len(incomingPaths)) // grab all files within all subfolders
		for _, incomingPath := range incomingPaths {
			if win.PathIsFolder(incomingPath) {
				nested, _ := win.EnumFilesDeep(incomingPath)
				allPaths = append(allPaths, nested...)
			} else {
				allPaths = append(allPaths, incomingPath)
			}
		}

		nonMp3Count := xslices.CountFunc(allPaths, func(_ int, path string) bool { // count how many non-MP3 we have
			return !win.PathHasExtension(path, "mp3")
		})
		if nonMp3Count == len(allPaths) { // zero MP3s found?
			me.wnd.UiThread(func() {
				me.wnd.Hwnd().MessageBox(
					fmt.Sprintf("No MP3 found amongst %d files.", len(allPaths)),
					"No MP3 files", co.MB_ICONERROR)
				me.setWaitState(false)
			})
			return // nothing do to
		}

		type TagAndPath struct { // a tag and its file path
			pTag *id3v2.Tag
			path string
		}

		mp3ToReadCount := len(allPaths) - nonMp3Count
		tags := make([]TagAndPath, 0, mp3ToReadCount) // load and cache all the MP3 tags
		for idxMp3, path := range allPaths {
			if win.PathHasExtension(path, "mp3") { // ignore non-MP3 files
				pTag, err := id3v2.TagFromFile(path)
				if err != nil {
					me.wnd.UiThread(func() {
						me.wnd.Hwnd().MessageBox(
							fmt.Sprintf("Error loading tag:\n%s\n\n%s", path, err.Error()),
							"Error", co.MB_ICONERROR)
						me.setWaitState(false)
					})
					return // stop on first error, no tag is loaded
				}
				tags = append(tags, TagAndPath{
					pTag: pTag,
					path: path,
				})

				me.wnd.UiThread(func() { // UI progress feedback
					me.wnd.Hwnd().SetWindowText(fmt.Sprintf("%d/%d files read...", idxMp3+1, mp3ToReadCount))
				})
			}
		}

		me.wnd.UiThread(func() { // finally fill the listview with the tags
			for _, tag := range tags {
				var item ui.ListViewItem
				if existingItem, ok := me.lstFiles.Items.Find(tag.path); ok { // file already loaded?
					item = existingItem // current tag object will be replaced
				} else {
					item = me.lstFiles.Items.AddWithIcon(0, tag.path) // insert new item
				}
				item.SetData(tag.pTag) // store tag in item
				me.renderMp3InList(item)
			}
			me.sortList()
			me.lstFiles.Cols.Get(0).SetWidthToFill()
			me.setWaitState(false)
		})
	}()
}

func (me *DlgMain) renderMp3InList(item ui.ListViewItem) {
	pTag := item.Data().(*id3v2.Tag) // retrieve tag stored in item
	if pTag.IsEmpty() {
		item.SetText(1, "N/A") // MP3 without tag
	} else {
		item.SetText(1, fmt.Sprintf("%d", pTag.Padding()))
	}

	if pTag.FrameByName4("APIC") != nil {
		item.SetText(2, "\u2713") // checkmark symbol
	} else {
		item.SetText(2, "")
	}

	item.SetText(3, pTag.ReplayGainStatus())

	me.renderMp3TextCell(item, 4, pTag, "TPE1")
	me.renderMp3TextCell(item, 5, pTag, "TYER")
	me.renderMp3TextCell(item, 6, pTag, "TALB")
	me.renderMp3TextCell(item, 7, pTag, "TRCK")
	me.renderMp3TextCell(item, 8, pTag, "TIT2")
	me.renderMp3TextCell(item, 9, pTag, "TCON")
	me.renderMp3TextCell(item, 10, pTag, "TPE3")
	me.renderMp3TextCell(item, 11, pTag, "TCOM")
	me.renderMp3TextCell(item, 12, pTag, "TEXT")
	me.renderMp3TextCell(item, 13, pTag, "TOPE")

	if pFrame := pTag.FrameByName4("COMM"); pFrame != nil {
		body, _ := pFrame.Body().(*id3v2.BodyComment)
		item.SetText(14, body.Text)
	} else {
		item.SetText(14, "") // clear
	}
}

func (me *DlgMain) renderMp3TextCell(item ui.ListViewItem, colIndex int, tag *id3v2.Tag, name4 string) {
	if pFrame := tag.FrameByName4(name4); pFrame != nil { // such name4 frame exists
		body, _ := pFrame.Body().(*id3v2.BodyText)
		item.SetText(colIndex, body.Text)
	} else {
		item.SetText(colIndex, "") // clear
	}
}

func (me *DlgMain) sortList() {
	me.lstFiles.Items.Sort(func(itemA, itemB ui.ListViewItem) int {
		cmp := 0
		if me.sortCol == 1 { // by padding size
			tagA := itemA.Data().(*id3v2.Tag)
			tagB := itemB.Data().(*id3v2.Tag)
			cmp = int(tagA.Padding()) - int(tagB.Padding())
		} else { // by column text
			cmp = wstr.CmpI(itemA.Text(me.sortCol), itemB.Text(me.sortCol))
		}

		if me.sortAsc {
			return cmp
		} else {
			return -cmp
		}
	})
}

func (me *DlgMain) removePicRg(delRg bool) bool {
	nFiles := me.lstFiles.Items.SelectedCount()
	text := fmt.Sprintf("Remove picture frame from %d file(s)?", nFiles)
	if delRg {
		text = fmt.Sprintf("Remove picture and ReplayGain frames from %d file(s)?", nFiles)
	}
	if ui.MsgOkCancel(me.wnd, "Remove frames", "", text, "&Remove") != co.ID_OK {
		return false
	}

	for _, item := range me.lstFiles.Items.Selected() {
		pTag := item.Data().(*id3v2.Tag)
		pTag.RemoveFrameIf(func(pFrame *id3v2.Frame) bool {
			if pFrame.Name4() == "APIC" {
				return true
			}
			if delRg && pFrame.Name4() == "TXXX" {
				if pBody, ok := pFrame.Body().(*id3v2.BodyUserText); ok {
					if strings.HasPrefix(pBody.Descr, "replaygain_track_") ||
						strings.HasPrefix(pBody.Descr, "replaygain_album_") {
						return true
					}
				}
			}
			return false
		})
	}
	return true
}

func (me *DlgMain) editSelected() bool {
	if me.lstFiles.Items.SelectedCount() == 0 {
		return false // Enter key will hit here even without selected items
	}

	clonedTags := xslices.Map(me.lstFiles.Items.Selected(), func(_ int, item ui.ListViewItem) *id3v2.Tag {
		pTag := item.Data().(*id3v2.Tag)
		return pTag.Clone()
	})

	if dlgedit.ShowNew(me.wnd, clonedTags) == co.ID_OK {
		for i, item := range me.lstFiles.Items.Selected() {
			item.SetData(clonedTags[i]) // replace the selected tags with the edited ones
		}
		return true
	}

	return false
}

func (me *DlgMain) saveSelectedAsync() {
	type (
		Failure struct { // a register of one saving error
			path string
			err  error
		}
		TagAndPath struct { // a tag and its file path
			pTag *id3v2.Tag
			path string
		}
	)

	failures := make([]Failure, 0) // to store all failures we get

	selTags := xslices.Map(me.lstFiles.Items.Selected(), func(_ int, item ui.ListViewItem) TagAndPath {
		return TagAndPath{
			pTag: item.Data().(*id3v2.Tag),
			path: item.Text(0),
		}
	})
	me.setWaitState(true)

	go func() {
		for idxTag, selTag := range selTags { // process each tag sequentially
			if err := selTag.pTag.SaveToFile(selTag.path); err != nil {
				failures = append(failures, Failure{selTag.path, err}) // store error, and keep going
			}

			me.wnd.UiThread(func() { // UI progress feedback
				me.wnd.Hwnd().SetWindowText(fmt.Sprintf("%d/%d files written...", idxTag+1, len(selTags)))
			})
		}

		me.wnd.UiThread(func() { // UI final feedback
			for _, item := range me.lstFiles.Items.Selected() {
				me.renderMp3InList(item) // re-render, all paddings have been removed
			}
			me.sortList()

			if len(failures) > 0 { // any errors?
				var sb strings.Builder
				sb.WriteString(fmt.Sprintf("%d file(s) failed to save:", len(failures)))
				for _, fail := range failures {
					sb.WriteString("\n\n")
					sb.WriteString(fail.path)
					sb.WriteString("\n")
					sb.WriteString(fail.err.Error())
				}
				ui.MsgError(me.wnd, "Error saving file(s)", "", sb.String())
			}
			me.setWaitState(false)
			me.lstFiles.Focus()
		})
	}()
}
